### PR TITLE
Laboratory work #4, Baranov Timofey - 22FPL1

### PR DESCRIPTION
**Pull request by tima-baranov**

### CODE REVIEW

**Review by artyomtugaryov — CHANGES_REQUESTED** [review]
Nothing to review

File: lab_4_fill_words_by_ngrams/main.py [hunk 1]
```diff
@@ -28,6 +30,18 @@ def _tokenize(self, text: str) -> tuple[str, ...]:  # type: ignore
         Raises:
             ValueError: In case of inappropriate type input argument or if input argument is empty.
         """
+        if not (isinstance(text, str) and text):
+            raise ValueError
```
Comment (artyomtugaryov) [inline lab_4_fill_words_by_ngrams/main.py:34]: add message

File: lab_4_fill_words_by_ngrams/main.py [hunk 2]
```diff
@@ -39,6 +53,12 @@ def _put(self, element: str) -> None:
         Raises:
             ValueError: In case of inappropriate type input argument or if input argument is empty.
         """
+        if not (isinstance(element, str) and element):
+            raise ValueError
```
Comment (artyomtugaryov) [inline lab_4_fill_words_by_ngrams/main.py:57]: .

File: lab_4_fill_words_by_ngrams/main.py [hunk 3]
```diff
@@ -39,6 +53,12 @@ def _put(self, element: str) -> None:
...
+        if element in self._storage:
+            return None
+        self._storage[element] = len(self._storage)
+        return None
```
Comment (artyomtugaryov) [inline lab_4_fill_words_by_ngrams/main.py:61]: There is a better solution. `return None` is ugly

File: lab_4_fill_words_by_ngrams/main.py [hunk 4]
```diff
@@ -56,6 +76,19 @@ def _postprocess_decoded_text(self, decoded_corpus: tuple[str, ...]) -> str:  #
         Raises:
             ValueError: In case of inappropriate type input argument or if input argument is empty.
         """
+        if not (isinstance(decoded_corpus, tuple)
+                and decoded_corpus):
+            raise ValueError
```
Comment (artyomtugaryov) [inline lab_4_fill_words_by_ngrams/main.py:81]: .

File: lab_4_fill_words_by_ngrams/main.py [hunk 5]
```diff
@@ -56,6 +76,19 @@ def _postprocess_decoded_text(self, decoded_corpus: tuple[str, ...]) -> str:  #
...
+        decoded_sentences = ''
+        list_of_tokens = list(decoded_corpus)
+        sentences = (' '.join(list_of_tokens)).split(self._end_of_word_token)
+        for sentence in sentences:
+            sentence = sentence.strip().capitalize()
+            if not sentence:
+                break
+            sentence = sentence + '. '
```
Comment (artyomtugaryov) [inline lab_4_fill_words_by_ngrams/main.py:89]: use fstring

File: lab_4_fill_words_by_ngrams/main.py [hunk 6]
```diff
@@ -98,6 +134,34 @@ def run(self, seq_len: int, prompt: str) -> str:  # type: ignore
                 or if sequence has inappropriate length,
                 or if methods used return None.
         """
+        if not isinstance(seq_len, int) or not isinstance(prompt, str) or not prompt or seq_len <= 0:
+            raise ValueError
```
Comment (artyomtugaryov) [inline lab_4_fill_words_by_ngrams/main.py:138]: .

File: lab_4_fill_words_by_ngrams/main.py [hunk 7]
```diff
@@ -98,6 +134,34 @@ def run(self, seq_len: int, prompt: str) -> str:  # type: ignore
...
+            for word, value in sorted_candidates:
+                probability += value
+                possible_tokens += (word,)
+                if probability >= self._p_value:
+                    random_word = random.choice(possible_tokens)
+                    encoded_text += (random_word,)
+                    break
+        decoded = self._word_processor.decode(encoded_text)
+        if decoded is None:
+            raise ValueError
```
Comment (artyomtugaryov) [inline lab_4_fill_words_by_ngrams/main.py:163]: .

**Review by artyomtugaryov — CHANGES_REQUESTED** [review]

File: lab_4_fill_words_by_ngrams/main.py (hunk 1 above)
Comment (tima-baranov) [reply lab_4_fill_words_by_ngrams/main.py]: fixed

File: lab_4_fill_words_by_ngrams/main.py (hunk 2 above)
Comment (tima-baranov) [reply lab_4_fill_words_by_ngrams/main.py]: fixed

File: lab_4_fill_words_by_ngrams/main.py (hunk 3 above)
Comment (tima-baranov) [reply lab_4_fill_words_by_ngrams/main.py]: fixed

File: lab_4_fill_words_by_ngrams/main.py (hunk 4 above)
Comment (tima-baranov) [reply lab_4_fill_words_by_ngrams/main.py]: fixed

File: lab_4_fill_words_by_ngrams/main.py (hunk 5 above)
Comment (tima-baranov) [reply lab_4_fill_words_by_ngrams/main.py]: fixed

File: lab_4_fill_words_by_ngrams/main.py (hunk 6 above)
Comment (tima-baranov) [reply lab_4_fill_words_by_ngrams/main.py]: fixed

File: lab_4_fill_words_by_ngrams/main.py (hunk 7 above)
Comment (tima-baranov) [reply lab_4_fill_words_by_ngrams/main.py]: fixed

**Comment by vitazaytseva** [comment]
Well done! Good job!

**Pull request closed without merging**